packages/snaps-execution-environments/src/offscreen/OffscreenSnapExecutor.ts

```diff
@@ -0,0 +1,166 @@
+import {
+  BasePostMessageStream,
+  WindowPostMessageStream,
+} from '@metamask/post-message-stream';
+import { JsonRpcParams, JsonRpcRequest } from '@metamask/utils';
+
+type ExecutorJob = {
+  id: string;
+  window: Window;
+  stream: WindowPostMessageStream;
+};
+
+export class OffscreenSnapExecutor {
+  readonly #stream: BasePostMessageStream;
+
+  readonly #jobs: Record<string, ExecutorJob> = {};
+
+  constructor(stream: BasePostMessageStream) {
+    this.#stream = stream;
+    this.#stream.on('data', this.#onData.bind(this));
+  }
+
+  /**
+   * Handle an incoming message from the `OffscreenExecutionService`. This
+   * assumes that the message contains a `jobId` property, and a JSON-RPC
+   * request in the `data` property.
+   *
+   * @param data - The message data.
+   * @param data.data - The JSON-RPC request.
+   * @param data.jobId - The job ID.
+   * @param data.frameUrl - The URL to load in the iframe.
+   */
+  #onData(data: {
+    data: JsonRpcRequest<JsonRpcParams>;
+    jobId: string;
+    frameUrl: string;
+  }) {
+    const { jobId, frameUrl, data: request } = data;
+
+    if (!this.#jobs[jobId]) {
+      // This ensures that a job is initialized before it is used. To avoid
+      // code duplication, we call the `#onData` method again, which will
+      // run the rest of the logic after initialization.
+      this.#initializeJob(jobId, frameUrl)
+        .then(() => {
+          this.#onData(data);
+        })
+        .catch((error) => {
+          console.error('[Worker] Error initializing job:', error);
+        });
+
+      return;
+    }
+
+    // This is a method specific to the `OffscreenSnapExecutor`, as the service
+    // itself does not have access to the iframes directly.
+    if (request.method === 'terminateJob') {
```

Comment:
The executor seems kind of hacky. Could we run the iframe execution environment in the offscreen doc executor instead or is that not possible?

Reply:
e.g. `SnapController <-> OffscreenExecutionService <-> IframeExecutionService (inside offscreen doc) <-> IframeSnapExecutor`

Reply:
Assuming you mean iframe execution _service_, it would be possible, but probably much more hacky. The service is deeply connected with our extension stack (controllers, Snaps provider, etc.), which would be difficult to replicate within the offscreen environment.

Reply:
Good point - hmm 🤔